metaseq/modules/multihead_attention.py

```diff
@@ -345,6 +365,17 @@ def forward(
                 )
 
         attn_weights = torch.bmm(q, k.transpose(1, 2))
+        if isinstance(attn_weights, ShardedTensor):
+            zero_mask = torch.ones(
+                *attn_weights.size(), device=attn_weights.local_tensor().device
+            )
+            row, col = attn_weights.size(1), attn_weights.size(2)
+            row //= world_size
+            col //= world_size
+            for i in range(world_size):
+                zero_mask[:, row * i : row * (i + 1), col * i : col * (i + 1)] = 0
+            zero_mask = zero_mask.type(torch.BoolTensor).cuda()
+            attn_weights = attn_weights.masked_fill(zero_mask, 0.0)
```

Comment:
Two questions:

1) Why do we need to do this since the original code doesn't have this logic?
2) `zero_mask` is a huge tensor and we are only using part of it for masked_fill, so maybe we can allocate only that much memory?

Reply:
Because we now gather input from different ranks and doing bmm will create interaction between weights between different ranks which are not needed. (They have no meaning, IMHO)

Reply:
And I agree, we can have a more efficient way to do this.